File: backend/src/coraClient.js
const CORA_API_URL = process.env.CORA_API_URL;
const CORA_API_CANTIDAD = Number(process.env.CORA_API_CANTIDAD ?? 24);

function normalizarLista(datos) {
  if (Array.isArray(datos)) return datos;
  if (Array.isArray(datos?.data)) return datos.data;
  if (Array.isArray(datos?.items)) return datos.items;
  if (Array.isArray(datos?.result)) return datos.result;
  if (datos && typeof datos === 'object') return [datos];
  return [];
}

function buscarValor(registro, candidatos) {
  if (!registro || typeof registro !== 'object') return null;

  const entradas = Object.entries(registro);
  for (const candidato of candidatos) {
    const exacto = entradas.find(([clave]) => clave.toLowerCase() === candidato);
    if (exacto) return exacto[1];
  }

  for (const candidato of candidatos) {
    const parcial = entradas.find(([clave]) => clave.toLowerCase().includes(candidato));
    if (parcial) return parcial[1];
  }

  return null;
}

function aNumero(valor) {
  if (valor == null || valor === '') return null;
  const numero = Number(String(valor).replace(',', '.'));
  return Number.isFinite(numero) ? numero : null;
}

function dosDigitos(valor) {
  return String(valor).padStart(2, '0');
}

function formatearFechaLocal(fecha) {
  return [
    fecha.getFullYear(),
    dosDigitos(fecha.getMonth() + 1),
    dosDigitos(fecha.getDate())
  ].join('-');
}

function formatearHoraLocal(fecha) {
  return [
    dosDigitos(fecha.getHours()),
    dosDigitos(fecha.getMinutes()),
    dosDigitos(fecha.getSeconds())
  ].join(':');
}

function separarFechaHora(registro) {
  const fechaLectura = buscarValor(registro, [
    'fechalectura',
    'fecha_lectura',
    'timestamp',
    'createdat',
    'created_at',
    'datetime',
    'fecha'
  ]);
  const horaLectura = buscarValor(registro, ['hora', 'time']);

  if (fechaLectura) {
    const fecha = new Date(fechaLectura);
    if (!Number.isNaN(fecha.getTime())) {
      return {
        fecha: formatearFechaLocal(fecha),
        hora: formatearHoraLocal(fecha)
      };
    }
  }

  return {
    fecha: String(fechaLectura ?? '').slice(0, 10),
    hora: String(horaLectura ?? '00:00:00').slice(0, 8)
  };
}

export function normalizarRegistroCora(registro) {
  const { fecha, hora } = separarFechaHora(registro);

  return {
    fecha,
    hora,
    nivel: aNumero(buscarValor(registro, ['nivel', 'embalse'])),
    qe: aNumero(buscarValor(registro, ['qe', 'caudalentrada', 'entrada', 'ingreso'])),
    qs: aNumero(buscarValor(registro, ['qs', 'caudalsalida', 'salida'])),
    qv: aNumero(buscarValor(registro, ['qv', 'caudalvertido', 'vertido', 'vertedero'])),
    potenciaActiva: aNumero(buscarValor(registro, [
      'potenciaactiva',
      'potencia_activa',
      'potencia',
      'pa',
      'mw'
    ])),
    clima: buscarValor(registro, ['clima', 'weather', 'condicion', 'condición']) ?? null,
    datosOriginales: registro
  };
}

export async function obtenerDatosCora() {
  if (!CORA_API_URL) {
    throw new Error('Falta configurar CORA_API_URL.');
  }

  const url = new URL(CORA_API_URL);
  url.searchParams.set('cantidad', String(CORA_API_CANTIDAD));

  const respuesta = await fetch(url, { cache: 'no-store' });
  if (!respuesta.ok) {
    throw new Error(`CORA respondió con estado ${respuesta.status}.`);
  }

  return normalizarLista(await respuesta.json())
    .map(normalizarRegistroCora)
    .filter(registro => registro.fecha && registro.hora);
}

File: backend/Clients/CoraClient.cs
using System.Globalization;
using System.Text.Json.Nodes;
using CaudalesBackend.Infrastructure;
using CaudalesBackend.Models;

namespace CaudalesBackend.Clients;

internal sealed class CoraClient
{
    private readonly HttpClient httpClient;

    public CoraClient(HttpClient httpClient)
    {
        this.httpClient = httpClient;
    }

    public async Task<List<DatoCora>> ObtenerDatosAsync(CoraOptions options)
    {
        var url = new UriBuilder(options.ApiUrl);
        var separator = string.IsNullOrWhiteSpace(url.Query) ? "" : "&";
        url.Query = $"{url.Query.TrimStart('?')}{separator}cantidad={options.Cantidad}";

        var json = await HttpJsonFetcher.GetStringAsync(httpClient, url.Uri);
        var root = JsonNode.Parse(json);

        return NormalizarLista(root)
            .Select(registro => NormalizarRegistro(options.Planta, registro))
            .Where(dato => dato is not null)
            .Cast<DatoCora>()
            .ToList();
    }

    private static IEnumerable<JsonNode> NormalizarLista(JsonNode? root)
    {
        if (root is JsonArray array) return array.OfType<JsonNode>();
        if (root?["data"] is JsonArray data) return data.OfType<JsonNode>();
        if (root?["items"] is JsonArray items) return items.OfType<JsonNode>();
        if (root?["result"] is JsonArray result) return result.OfType<JsonNode>();
        return root is null ? [] : [root];
    }

    private static DatoCora? NormalizarRegistro(string planta, JsonNode registro)
    {
        var (fecha, hora) = SepararFechaHora(registro);
        if (fecha is null || hora is null) return null;

        return new DatoCora(
            planta,
            fecha.Value,
            hora.Value,
            BuscarDecimal(registro, ["nivel", "embalse"]),
            BuscarDecimal(registro, ["qe", "caudalentrada", "entrada", "ingreso"]),
            BuscarDecimal(registro, ["qs", "caudalsalida", "salida"]),
            BuscarDecimal(registro, ["qv", "caudalvertido", "vertido", "vertedero"]),
            BuscarDecimal(registro, ["potenciaactiva", "potencia_activa", "potencia", "pa", "mw"]),
            BuscarTexto(registro, ["clima", "weather", "condicion", "condición"]),
            registro.DeepClone());
    }

    private static (DateOnly? Fecha, TimeOnly? Hora) SepararFechaHora(JsonNode registro)
    {
        var fechaLectura = BuscarTexto(registro, ["fechalectura", "fecha_lectura", "timestamp", "createdat", "created_at", "datetime", "fecha"]);
        var horaLectura = BuscarTexto(registro, ["hora", "time"]);

        if (DateTime.TryParse(fechaLectura, CultureInfo.InvariantCulture, DateTimeStyles.AssumeLocal, out var fechaHora))
        {
            return (DateOnly.FromDateTime(fechaHora), TimeOnly.FromDateTime(fechaHora));
        }

        var fecha = DateOnly.TryParse(fechaLectura, out var soloFecha) ? soloFecha : (DateOnly?)null;
        var hora = TimeOnly.TryParse(horaLectura, out var soloHora) ? soloHora : new TimeOnly(0, 0);
        return (fecha, hora);
    }

    private static string? BuscarTexto(JsonNode registro, string[] candidatos)
    {
        if (registro is not JsonObject obj) return null;

        foreach (var candidato in candidatos)
        {
            var exacto = obj.FirstOrDefault(prop => string.Equals(prop.Key, candidato, StringComparison.OrdinalIgnoreCase));
            if (exacto.Value is not null) return exacto.Value.ToString();
        }

        foreach (var candidato in candidatos)
        {
            var parcial = obj.FirstOrDefault(prop => prop.Key.Contains(candidato, StringComparison.OrdinalIgnoreCase));
            if (parcial.Value is not null) return parcial.Value.ToString();
        }

        return null;
    }

    private static decimal? BuscarDecimal(JsonNode registro, string[] candidatos)
    {
        var texto = BuscarTexto(registro, candidatos);
        if (string.IsNullOrWhiteSpace(texto)) return null;

        texto = texto.Replace(',', '.');
        return decimal.TryParse(texto, NumberStyles.Any, CultureInfo.InvariantCulture, out var numero)
            ? numero
            : null;
    }
}
